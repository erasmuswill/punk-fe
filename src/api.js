import swr, { useSWRInfinite } from "swr";
import axios from "axios";
import { useContext } from "react";
import { ShowAlcoholicContext } from "./App";
const queryString = require("query-string");

axios.defaults.baseURL = "https://api.punkapi.com/v2";

let cancelSource = axios.CancelToken.source();

const requestInterceptor = (config) => {
  config.cancelToken = cancelSource.token;
  return config;
};

axios.interceptors.request.use(requestInterceptor);

const fetcher = (url) =>
  axios.get(url).then(({ data }) => {
    data = data.filter(
      ({ description, image_url }) => description && image_url
    );
    if (data.length) return data;
    return fetcher(url);
  });
const minorFetcher = (url) =>
  axios.get(url).then(({ data }) => {
    data = data.filter(
      ({ abv, description, image_url }) =>
        abv <= 0.5 && description && image_url
    );
    if (data.length) return data;
    console.log("retrying");
    return minorFetcher(url);
  });

function useSWR(url, isIdempotent) {
  const showAlcoholic = useContext(ShowAlcoholicContext);
  let config = {
    suspense: true,
    loadingTimeout: 1500,
    fetcher: !showAlcoholic && url === "/beers/random" ? minorFetcher : fetcher,
  };
  if (!isIdempotent)
    config = {
      ...config,
      dedupingInterval: 100,
      revalidateOnFocus: false,
      revalidateOnReconnect: false,
    };
  return swr(url, config);
}

function firstIndex(arr) {
  if (arr && Array.isArray(arr) && arr.length) return arr[0];
  else return undefined;
}

function useBeer(id) {
  const { data, error, isValidating, mutate } = useSWR(
    id ? `/beers/${id}` : null
  );

  return {
    beer: data[0],
    isLoading: !error && !data,
    isError: error,
  };
}

function useRandomBeer() {
  const { data, error, isValidating, mutate } = useSWR("/beers/random");

  return {
    beer: firstIndex(data),
    isLoading: !error && !data,
    isError: error,
    isValidating,
    mutate,
  };
}

function useBeerList(filter = []) {
  const showAlcoholic = useContext(ShowAlcoholicContext);
  console.log({ filter });

  const filterObj = {};
  if (filter.length) Object.assign(filterObj, ...filter);

  let { data, error, size, setSize, isValidating } = useSWRInfinite(
    (index) =>
      `/beers?${queryString.stringify({ ...filterObj, page: index + 1 })}`,
    {
      fetcher: (url) =>
        axios.get(url).then(({ data }) => {
          data = data.filter(
            ({ description, image_url }) => description && image_url
          );
          return data;
        }),
      suspense: true,
      loadingTimeout: 1500,
      dedupingInterval: 100,
      revalidateOnFocus: false,
      revalidateOnReconnect: false,
    }
  );
  const isLoadingInitialData = !data && !error,
    isLoadingMore =
      isLoadingInitialData ||
      (size > 0 && data && typeof data[size - 1] === "undefined"),
    hasNextPage = data[data.length - 1].length;

  data = data.filter((v) => v.length).flat();

  if (!showAlcoholic) {
    data = data.filter(({ abv }) => abv <= 0.5);
    if (data.length < 10 && hasNextPage && !isLoadingMore) setSize(size + 1);
  }

  return {
    beers: data,
    isLoadingMore,
    isLoadingInitialData,
    isError: error,
    size,
    setSize,
    isValidating,
    hasNextPage,
  };
}

export { useBeer, useRandomBeer, useBeerList };
