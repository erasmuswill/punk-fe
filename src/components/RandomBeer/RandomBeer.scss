div.random-beer {
  min-height: 212px;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: space-between;
  align-content: center;
  align-items: center;
  border-radius: 12px;
  border: 1px solid #1890ff !important;
  padding: 16px;
  margin: 22px;

  div.title-block {
    flex-basis: 10%;
    img {
      margin: 8px auto;
      display: block;
    }
  }
  div.label-block {
    flex-basis: 60%;
    margin: 0px 24px;
  }
  div.cta-block {
    flex-basis: 20%;
    button {
      width: 150px;
    }
    button:first-of-type {
      margin-bottom: 16px;
    }
  }
}
