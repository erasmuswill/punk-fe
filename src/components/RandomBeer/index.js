import { Button, Tooltip, Typography } from "antd";
import { ReloadOutlined } from "@ant-design/icons";
import { useState } from "react";
import { useRandomBeer } from "../../api";
import BeerDrawer from "../BeerDrawer";
import BeerItem from "../BeerItem";

const { Title } = Typography;

export default function RandomBeer() {
  const { beer = {}, loading, isValidating, mutate } = useRandomBeer();
  const [modalId, setModalId] = useState(null);
  return (
    <>
      <Title level={3}>I'm feeling lucky</Title>
      <BeerItem {...beer} setModalId={setModalId} />
      <Button
        type="primary"
        disabled={loading || isValidating}
        onClick={() => mutate()}
        icon={
          <ReloadOutlined
            spin={isValidating}
          />
        }
      >
        New random beer
      </Button>
      <BeerDrawer id={modalId} close={() => setModalId(null)} />
    </>
  );
}
