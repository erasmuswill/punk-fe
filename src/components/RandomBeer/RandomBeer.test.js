import { fireEvent, render } from "@testing-library/react";
import * as api from "../../api";
import RandomBeer from "./";
import TEST_DATA from '../../TEST_DATA.json'

describe("RandomBeer", () => {
  it("shows a random beer", () => {
    const mockMutate = jest.fn();
    jest.spyOn(api, "useRandomBeer").mockImplementation(() => ({
      beer: TEST_DATA.alc_beer,
      loading: false,
      isValidating: false,
      mutate: mockMutate,
    }));
    const { asFragment } = render(<RandomBeer />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("allows a new random beer to be shown", () => {
    const mockMutate = jest.fn();
    jest.spyOn(api, "useRandomBeer").mockImplementation(() => ({
      beer: TEST_DATA.alc_beer,
      loading: false,
      isValidating: false,
      mutate: mockMutate,
    }));
    const { getByText } = render(<RandomBeer />);
    const mutateButton = getByText("Random beer").parentElement; // This matches the span inside the button, calling parentElement to get a reference to the button
    fireEvent.click(mutateButton);
    expect(mockMutate).toHaveBeenCalledTimes(1);
  });

  it("shows a validation state", () => {
    const mockMutate = jest.fn();
    jest.spyOn(api, "useRandomBeer").mockImplementation(() => ({
      beer: undefined,
      loading: true,
      isValidating: true,
      mutate: mockMutate,
    }));
    const { asFragment } = render(<RandomBeer />);
    expect(asFragment()).toMatchSnapshot();
  });
});
