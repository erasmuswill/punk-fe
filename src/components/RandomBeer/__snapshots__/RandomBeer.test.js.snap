// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomBeer shows a random beer 1`] = `
<DocumentFragment>
  <h3
    class="ant-typography"
  >
    I'm feeling lucky
  </h3>
  <div
    class="beer-item"
  >
    <div
      class="image"
    >
      <img
        alt="Label for Punk IPA 2007 - 2010"
        height="100"
        src="https://images.punkapi.com/v2/192.png"
      />
    </div>
    <div
      class="name"
    >
      <h3>
        Punk IPA 2007 - 2010
      </h3>
    </div>
    <span
      class="abv"
    >
      6% alcohol
    </span>
    <div
      class="description"
    >
      <div
        class="display-text-group"
      >
        <span
          class="displayed-text"
        >
          Our flagship beer that kick started the craft beer revolution. This is James and Martin's original take on an American IPA, subverted with punchy New Zealand hops. Layered with new world hops to create an all-out riot of grapefruit, pineapple and lychee before a spiky, mouth-puckering bitter finish. 
        </span>
      </div>
    </div>
  </div>
  <button
    class="ant-btn ant-btn-primary"
    type="button"
  >
    <span
      aria-label="reload"
      class="anticon anticon-reload"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="reload"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M909.1 209.3l-56.4 44.1C775.8 155.1 656.2 92 521.9 92 290 92 102.3 279.5 102 511.5 101.7 743.7 289.8 932 521.9 932c181.3 0 335.8-115 394.6-276.1 1.5-4.2-.7-8.9-4.9-10.3l-56.7-19.5a8 8 0 00-10.1 4.8c-1.8 5-3.8 10-5.9 14.9-17.3 41-42.1 77.8-73.7 109.4A344.77 344.77 0 01655.9 829c-42.3 17.9-87.4 27-133.8 27-46.5 0-91.5-9.1-133.8-27A341.5 341.5 0 01279 755.2a342.16 342.16 0 01-73.7-109.4c-17.9-42.4-27-87.4-27-133.9s9.1-91.5 27-133.9c17.3-41 42.1-77.8 73.7-109.4 31.6-31.6 68.4-56.4 109.3-73.8 42.3-17.9 87.4-27 133.8-27 46.5 0 91.5 9.1 133.8 27a341.5 341.5 0 01109.3 73.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 003 14.1l175.6 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c-.1-6.6-7.8-10.3-13-6.2z"
        />
      </svg>
    </span>
    <span>
      New random beer
    </span>
  </button>
</DocumentFragment>
`;

exports[`RandomBeer shows a validation state 1`] = `
<DocumentFragment>
  <h3
    class="ant-typography"
  >
    I'm feeling lucky
  </h3>
  <button
    class="ant-btn ant-btn-primary"
    disabled=""
    type="button"
  >
    <span
      aria-label="reload"
      class="anticon anticon-reload anticon-spin"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="reload"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M909.1 209.3l-56.4 44.1C775.8 155.1 656.2 92 521.9 92 290 92 102.3 279.5 102 511.5 101.7 743.7 289.8 932 521.9 932c181.3 0 335.8-115 394.6-276.1 1.5-4.2-.7-8.9-4.9-10.3l-56.7-19.5a8 8 0 00-10.1 4.8c-1.8 5-3.8 10-5.9 14.9-17.3 41-42.1 77.8-73.7 109.4A344.77 344.77 0 01655.9 829c-42.3 17.9-87.4 27-133.8 27-46.5 0-91.5-9.1-133.8-27A341.5 341.5 0 01279 755.2a342.16 342.16 0 01-73.7-109.4c-17.9-42.4-27-87.4-27-133.9s9.1-91.5 27-133.9c17.3-41 42.1-77.8 73.7-109.4 31.6-31.6 68.4-56.4 109.3-73.8 42.3-17.9 87.4-27 133.8-27 46.5 0 91.5 9.1 133.8 27a341.5 341.5 0 01109.3 73.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 003 14.1l175.6 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c-.1-6.6-7.8-10.3-13-6.2z"
        />
      </svg>
    </span>
    <span>
      New random beer
    </span>
  </button>
</DocumentFragment>
`;
